modules/basis/basis-datasource/tomcat-listener/src/main/java/io/dirigible/tomcat/listener/DelegatingObjectFactory.java

```
@@ -0,0 +1,141 @@
+        Object objectInstance = null;
+        try {
+          objectInstance = factory.getObjectInstance(obj, name, nameCtx, environment);
+        } catch (NameNotFoundException nne) {
+          LOGGER.fine("Will move to next factory as facroty [" + factory + "] could not retrieve object instance for name [" + name
+              + "]. NameNotFoundException was thrown. " + nne.getMessage());
+        } catch (Exception e) {
+          LOGGER.severe("Exception is thrown by facroty [" + factory + "] during retrieving object instance. Exception is: " + e);
+          NamingException ne = new NamingException("Cannot create resource  object instance due to exception in the object factory");
+          ne.initCause(e);
+          throw ne;
+        }
+        if (objectInstance != null) {
+          LOGGER.fine("Object instance with name [" + name + "]is created by factory [" + factory + "].");
+          return objectInstance;
+        }
+      }
+    } else {
+      LOGGER.fine("No object factories are found for referenceType [" + referenceType + "].");
+    }
+    return null;
+  }
+
+  private Object getObjectInstanceFromDefaults(Object obj, Name name, Context nameCtx, Hashtable<?, ?> environment) throws NamingException {
+    final Reference reference = (Reference) obj;
+    RefAddr removedFactory = removeDelegatingObjectFactoryFromReference(reference);
+
+    Object objectInstance = null;
+
+    try {
+      String factoryClassName = reference.getFactoryClassName();
+      if (factoryClassName != null) {
+        ObjectFactory objectFactory = (ObjectFactory) getClass().getClassLoader().loadClass(factoryClassName).newInstance();
+        objectInstance = objectFactory.getObjectInstance(reference, name, nameCtx, environment);
+      }
+    } catch (NameNotFoundException nne) {
+      LOGGER.fine("Default object creation factory could not retrieve object instance for name [" + name
+          + "]. NameNotFoundException was thrown. " + nne.getMessage());
```

Shouldn't we also re-throw the exception here?

---

No this is part of the logic of resource creation